api/envoy/extensions/load_balancing_policies/least_request/v3/least_request.proto

```diff
@@ -22,10 +22,17 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // This configuration allows the built-in LEAST_REQUEST LB policy to be configured via the LB policy
 // extension point. See the :ref:`load balancing architecture overview
 // <arch_overview_load_balancing_types>` for more information.
-// [#next-free-field: 6]
 message LeastRequest {
+  // Available methods for selecting the host set from which to return the host with the
+  // fewest active requests.
+  enum SelectionMethod {
+    POWER_OF_N_CHOICES = 0;
```

Comment:
Instead of making `POWER_OF_N_CHOICES` be value 0, I suggest having value 0 be `UNSPECIFIED`, and then having the two choices you want as 1 and 2.  This way, control planes can either set the value to an explicit choice, or they can let it default to `UNSPECIFIED`, in which case the data plane can assume a reasonable default.

Reply:
Wouldn't that open us up to potential inconsistencies if the same config is consumed by multiple data planes? IMO we should be opinionated about the default selection method here, since the least request LB has always worked this way. We should only use a full scan selection method if it's explicitly chosen because it results in a thundering herd in all but a few special scenarios.

Reply:
We don't strictly need this enum if we also make `choice_count >= num_hosts` do the full scan?  You can set `choice_count=MAX_UINT32` to get this behavior.

Or do we anticipate other potential selection methods?  If so, it might be nicer to do something like:

```proto
oneof selection_method {
	FullScan full_scan = 5;
	NChoices n_choices = 6;
}
message FullScan {} // no config.
message NChoices {
	uint32 choice_count = 1;
}

google.protobuf.UInt32Value choice_count = 1 [deprecated=true]; // use n_choices.choice_count instead.
```

Reply:
FWIW, note that the [xDS API style guide](https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md) says not to use `oneof`, because it makes it hard to add new alternatives in a backward-compatible way.

Reply:
We also cannot have `choice_count >= num_hosts` do the full scan because this subtly changes the load balancing behavior and causes issues. The original issue tied to this PR has links to the discussions about the specifics.

Reply:
> We also cannot have choice_count >= num_hosts do the full scan

Isn't that what this PR does? I see the text in the PR description "This approach preserves automatically using full scan mode when choice_count > number_of_hosts" and "If set to FULL_SCAN, or if choice_count is greater than or equal to the number of hosts,..."

> The original issue tied to this PR has links to the discussions about the specifics.

Where? I didn't see anything in #11004. The only issue I see is a mention in this PR that you need to randomize the starting point for the scan.

Reply:
> That functionality was previously removed in #30794 because it caused the first host to always be selected if cluster stats were disabled. However, that justification no longer applies

Actually, if some existing implementations have an issue with setting `choice_count=MAX_UINT32` then we probably shouldn't be recommending this as a convention.  And if `oneof` is not allowed, then I suppose you can just disregard my previous comment entirely.

So in that case, I agree that having the default be `N_CHOICES` (bikeshedding notwithstanding) would be the ideal behavior.  If we wanted to add an enum value for `CONTROL_PLANE_CHOOSES` (i.e. @markdroth's `UNSPECIFIED`) that is _not_ the default but doesn't specify a behavior, then that seems reasonable to me if such a thing might be useful.

Reply:
> These differences are larger for small numbers of hosts, which is the only regime where a full scan selection is feasible.

@tonya11en : A digression / concern about large numbers of hosts:

One of my concerns is that, unlike the original contributor, my primary use case involves a large number of hosts. I effectively need "return a random host that has 0 active connections", which least request load balancer with full scan mode can accomplish (though inefficiently). I just don't know yet whether it will be _too_ inefficient with a large number of hosts (assuming connections are long-lived). Or whether N_CHOICES with 2 << N << len(hosts) would actually be better for such a use case (with the expectation that you're likely to find a host with 0 connections in N choices most of the time with large enough N).

Reply:
Ah, well for your use case I wouldn't recommend using FULL_SCAN @jkirschner-hashicorp . You'd be better served by setting the choice count to be the highest number you'd be willing to tolerate performance-wise. This way, if the performance of a full scan for 1000 hosts is the most you would be willing to tolerate, it puts an upper-bound on how many hosts the LB will scan through. 

The new behavior that this FULL_SCAN patch is actually introducing is that it cannot choose the same host twice when evaluating the least requests. For most LB deployments, this is going to cause thundering herds any time new/idle hosts are introduced into a pool of "busy" hosts, like during autoscaling events... Imagine having 100 envoys sending requests to a fleet of hosts and scaling up- all 100 envoys are going to send their requests to the new hosts and potentially overload them. 

There's [an envoy blog post](https://blog.envoyproxy.io/examining-load-balancing-algorithms-with-envoy-1be643ea121c) that talks about this in more detail, but imagine you have a scenario with 1 idle host among 10 busy ones:

If you take the FULL_SCAN approach, selection probabilities will heavily skew towards the new host, potentially causing too much traffic to go that way.
![image](https://github.com/envoyproxy/envoy/assets/1109246/9af958a9-36ef-4485-acb8-6bd9be10cbee)

Using the n-choice selection, we just _bias_ toward the new host- it receives more traffic than everyone else, but it's not an avalanche.
![image](https://github.com/envoyproxy/envoy/assets/1109246/7d46c4b9-20b0-4ddc-971e-e19b50af4485)

In the n-choice scenario, increasing the number of choices will cause the distribution to skew more heavily towards the idle (red in graph) node, but also be more resilient to thundering herds than a full scan.

Reply:
@jkirschner-hashicorp did you catch the comment above?

Reply:
Yes. In my particular use case, there will always be a large number of hosts with 0 active requests (assuming autoscaling is working reasonably well), so I'm not concerned about a thundering herd---as long as the pick ratio between the hosts tied for "least" (0 active requests) is reasonable rather than concentrated heavily (like always happens with a fixed start index, and happens sometimes with a random start index as you previously pointed out).

That said, in the short-term, I'm going to use N choices with a sufficiently large N because it's something that is already available.